### ISSUE TITLE
SSA Apply: Different or same command?

### ISSUE THREAD

**Issue by apelisse** [kubernetes/kubernetes#91269]
I've had different form of this conversation with many different people and so far we've been going toward making server-side apply the same command, but there are pros and cons to both approach. I'd like to have a more formal discussion about this.

I'm going to try to describe the problem neutrally, but I think I'm biased toward having a separate command.

Why should we have a new `kubectl` command:

- We're less likely to break anyone's workflow (flags not working as expected, something that used to work with `kubectl apply` doesn't work anymore), we already have a few of these and there are probably a lot that we haven't seen:
  - `--force` flag doesn't work for server-side apply
  - Missing protocol doesn't work on deployment ports.
  - Conflicts will start causing failures where they previously didn't
- The flags are a little messy because some clearly apply to client-side while other clearly apply to server-side, that's general engineer malpractice to convolute these two things.
- Because of the two points above, we're losing freedom to make the changes that we would want to make because we sort of have to maintain compatibility (C++ syndrome).
- Basically, the abstraction is leaky in bad ways.

On the other hand, re-using the same name is great because:
- There is clearly no better name than `apply` for the command, so we'll have to settle with something not as good
- If we ever want to re-claim the `apply` name, we'll need a long-term plan to deprecate the existing `apply` and eventually rename
- People will start using it instantly. The risk with the new command is that people never start using it or aren't aware that it exists

**Comment by apelisse** [comment]
cc @lavalamp @kwiesmueller @justinsb @seans3 @kubernetes/sig-cli-maintainers 
/wg api-expression
/sig cli
/sig api-machinery

**Comment by lavalamp** [comment]
> Missing protocol doesn't work on deployment ports

We will fix this soon, no? Before flipping over the command?

> Conflicts will start causing failures where they previously didn't

IMO this is a good thing, it wasn't doing what users thought it was previously.

> `--force` flag doesn't work for server-side apply

The `--force` flag should be removed / renamed. (But there's no reason why it couldn't work for SSA, just delete and then do the apply operation.)

> The flags are a little messy because some clearly apply to client-side while other clearly apply to server-side, that's general engineer malpractice to convolute these two things.

A list of flags would be helpful for evaulating how bad this is; this same sentence could apply to problems of very different severity :)

> Because of the two points above, we're losing freedom to make the changes that we would want to make because we sort of have to maintain compatibility (C++ syndrome).

We can break current users once or make future users' lives miserable.

If we do one break, I think on balance it's net good. If we break things all the time, on balance it's definitely net bad.

**Comment by justinsb** [comment]
I am personally wary of changing the behaviour of apply, even if we believe that the new behaviour is better.

If we're going to do so, a list of the different behaviours (including the good ones which justify the change) would be really helpful, I think.

For example, I believe there's a different behaviour when we remove a field from the source yaml and re-apply.

**Comment by apelisse** [comment]
@MikeSpreitzer

**Comment by lavalamp** [comment]
I believe the stakeholders on this are @kubernetes/sig-cli-maintainers. e.g. @seans3, @pwittrock, @soltysh, can one of you weight in?

I'd like to get clarity on this soon.

Personally, I think it's quite important to not offer users a plethora of apply choices. It'd be super confusing.

> If we're going to do so, a list of the different behaviours (including the good ones which justify the change) would be really helpful, I think.

I agree.

> For example, I believe there's a different behaviour when we remove a field from the source yaml and re-apply.

There's a couple differences, one is IMO a bug and we can't go on-by-default without fixing it, but the other is WAI, fixing CSA bugs. I'll continue on this topic in the doc Antoine started.
